# Existing Estimators Research

The goal of this notebook is to evaluate a number of existing password estimators, with the following goals:

1. Establish a baseline of how well they work against a pre-defined dataset
2. Understand how well they can handle the corner cases (i.e. passwords that appear to be secure, but aren't, and the other way around)

We start with 5 arbitrary reasonably popular libraries.

In [1]:
# Let's start by installing the required libraries
!pip install zxcvbn password_strength passwordmeter password-validator

# Following libraries we'll need for the actual evaluation
!pip install datasets scikit-learn ipywidgets numpy pandas tqdm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.9/51.9 kB 2.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 409.4/409.4 kB 8.3 MB/s eta 0:00:00
  Created wheel for passwordmeter: filename=passwordmeter-0.1.8-py3-none-any.whl size=48341 sha256=1a007042b358b3f447fce85fd15e56f4bee5561ec28fe69225aa88d0e59594b9
  Stored in directory: /root/.cache/pip/wheels/fb/7b/d5/6f48559036041839f76322f6a88c0464b9c447daea82c363dc
  Created wheel for asset: filename=asset-0.6.13-py3-none-any.whl size=30705 sha256=8ea96684e8457447d284706131273d76520b67a5ff64bfa32b07f13cda83e35d
  Stored in directory: /root/.cache/pip/wheels/0d/aa/fd/2e57366f68d75e6da6c68a0b5ac8c28dd452a15842b0e5a172
  Created wheel for aadict: filename=aadict-0.2.3-py3-none-any.whl size=16569 sha256=d50c0843cdd85f3822e5f4e33f1fcb56e7649a6e5b4e50a2f0a0085c81da3e6b

We will now write a method which would allow us to validate password strength using all 4 libraries

In [2]:
def check_password(password, method="zxcvbn"):
    """
    Check the strength of a password using one of several popular Python libraries.

    The following methods are available:
      - "zxcvbn": Uses the zxcvbn-python library. Secure if score >= 3.
      - "password_validator": Uses the password-validator package with a defined schema.
      - "password_strength": Uses the password_strength package and evaluates entropy.
                               Secure if entropy > 50 bits.
      - "passwordmeter": Uses the passwordmeter package. Secure if score > 0.5.
      - "password_checker": Uses the py-password-checker package (assumed API).
                            Secure if the checker validates the password.

    Args:
        password (str): The password to be evaluated.
        method (str): Which method to use for estimation.

    Returns:
        int: 1 if the password is considered secure, 0 otherwise.
    """

    if password is None:
        return 0

    if method.lower() == "zxcvbn":
        try:
            from zxcvbn import zxcvbn
            result = zxcvbn(password)
            # zxcvbn scores range from 0 (weak) to 4 (strong)
            return 1 if result.get("score", 0) >= 3 else 0
        except ImportError:
            print("Error: zxcvbn library is not installed.")
            return 0

    elif method.lower() == "password_validator":
        try:
            from password_validator import PasswordValidator
            # Define a schema: minimum 8 characters, maximum 100, at least one uppercase, one lowercase, and one digit.
            schema = PasswordValidator()
            schema.min(8).max(100).has().uppercase().has().lowercase().has().digits()
            is_valid = schema.validate(password)
            return 1 if is_valid else 0
        except ImportError:
            print("Error: password-validator module is not installed.")
            return 0
        except Exception as e:
            print(f"password_validator check failed: {e}")
            return 0

    elif method.lower() == "password_strength":
        try:
            from password_strength import PasswordStats
            stats = PasswordStats(password)

            # Documentation suggests 0.66 as a good magic number
            return 1 if stats.strength() > 0.66 else 0
        except ImportError:
            print("Error: password_strength module is not installed.")
            return 0
        except Exception as e:
            print(f"password_strength check failed: {e}")
            return 0

    elif method.lower() == "passwordmeter":
        try:
            import passwordmeter
            score, improvements = passwordmeter.test(password)
            # Assume a score greater than 0.5 indicates a secure password.
            return 1 if score > 0.5 else 0
        except ImportError:
            print("Error: passwordmeter module is not installed.")
            return 0
        except Exception as e:
            print(f"passwordmeter check failed: {e}")
            return 0

    else:
        raise ValueError("Unknown method specified. Choose one of: zxcvbn, password_validator, password_strength, passwordmeter, password_checker.")


# Example usage:
test_password = "ExamplePass123!"
VALIDATION_METHODS = ["zxcvbn", "password_validator", "password_strength", "passwordmeter"]

for m in VALIDATION_METHODS:
    result = check_password(test_password, method=m)
    print(f"Method '{m}': {'Secure' if result == 1 else 'Not secure'}")

Method 'zxcvbn': Secure
Method 'password_validator': Secure
Method 'password_strength': Secure
Method 'passwordmeter': Secure


We can now load a dataset which we'll be using for evaluation. We'll use `InfinitodeLTD/PWLDS` which is a very large synthetic dataset created for this very purpose.

In [6]:
import random
from datasets import load_dataset

ds = load_dataset("InfinitodeLTD/PWLDS", split="train")

# Let's only take a subset of the data for faster processing
# Value 2 corresponds to neither secure nor insecure passwords, so we'll exclude it
my_filter = lambda x: random.random() < 0.01 and x["Strength_Level"] != 2

ds = ds.filter(my_filter)

Filter:   0%|          | 0/10000192 [00:00<?, ? examples/s]

Let's jot down some general information about the dataset:

In [7]:
print("General Dataset Information:")
print("----------------------------")
print("Total examples:", ds.num_rows)
print("Columns:", ds.column_names)
print("Features:", ds.features)
print("\nSample entries:")
for i in range(min(5, ds.num_rows)):
    print(ds[i])

General Dataset Information:
----------------------------
Total examples: 80140
Columns: ['Password', 'Strength_Level']
Features: {'Password': Value(dtype='string', id=None), 'Strength_Level': Value(dtype='int64', id=None)}

Sample entries:
{'Password': 'qmpo', 'Strength_Level': 0}
{'Password': 'eyju', 'Strength_Level': 0}
{'Password': 'ssgsh', 'Strength_Level': 0}
{'Password': 'arrisc', 'Strength_Level': 0}
{'Password': '5ukjv', 'Strength_Level': 0}


Cool, let's run this dataset through our estimators:

In [10]:
from sklearn.metrics import confusion_matrix, classification_report
from tqdm.notebook import tqdm
import pandas as pd
from IPython.display import display, Markdown

def map_label(strength):
    return 1 if strength in [3, 4] else 0

# Define true_labels based on your dataset's 'Strength_Level'
true_labels = [map_label(sl) for sl in ds["Strength_Level"]]

for method in VALIDATION_METHODS:
    predictions = [
        check_password(password, method=method)
        for password in tqdm(ds["Password"], desc=f"Processing with {method}")
    ]

    # Compute confusion matrix and classification report
    cm = confusion_matrix(true_labels, predictions)
    cr = classification_report(true_labels, predictions, output_dict=True)

    # Format Markdown output
    md_output = f"## Evaluating Method: **{method}**\n\n"

    # Confusion Matrix
    cm_df = pd.DataFrame(
        cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
    )
    md_output += "**Confusion Matrix:**\n\n"
    md_output += cm_df.to_markdown() + "\n\n"

    # Classification Report
    cr_df = pd.DataFrame(cr).transpose().round(2)
    cr_df.index = cr_df.index.map(lambda x: {"0": "Class 0", "1": "Class 1"}.get(x, x.title()))
    md_output += "**Classification Report:**\n\n"
    md_output += cr_df.to_markdown() + "\n\n"

    # Identify mispredictions (up to 5 FPs and 10 FNs)
    mispred_indices = [i for i, (t, p) in enumerate(zip(true_labels, predictions)) if t != p]
    fp_indices = [i for i in mispred_indices if true_labels[i] == 0 and predictions[i] == 1][:5]
    fn_indices = [i for i in mispred_indices if true_labels[i] == 1 and predictions[i] == 0][:10]

    md_output += "**Mispredictions (up to 5 FPs and 10 FNs):**\n\n"

    if not (fp_indices or fn_indices):
        md_output += "No mispredictions found!\n"
    else:
        mispred_list = []
        for idx in fp_indices:
            mispred_list.append({
                "Type": "False Positive",
                "Password": f"`{ds['Password'][idx]}`",
                "True Label": true_labels[idx],
                "Prediction": predictions[idx]
            })
        for idx in fn_indices:
            mispred_list.append({
                "Type": "False Negative",
                "Password": f"`{ds['Password'][idx]}`",
                "True Label": true_labels[idx],
                "Prediction": predictions[idx]
            })

        mispred_df = pd.DataFrame(mispred_list)
        md_output += mispred_df.to_markdown(index=False)

    # Display formatted markdown
    display(Markdown(md_output))

Processing with zxcvbn:   0%|          | 0/80140 [00:00<?, ?it/s]

## Evaluating Method: **zxcvbn**

**Confusion Matrix:**

|          |   Predicted 0 |   Predicted 1 |
|:---------|--------------:|--------------:|
| Actual 0 |         24101 |         15887 |
| Actual 1 |          1732 |         38420 |

**Classification Report:**

|              |   precision |   recall |   f1-score |   support |
|:-------------|------------:|---------:|-----------:|----------:|
| Class 0      |        0.93 |     0.6  |       0.73 |  39988    |
| Class 1      |        0.71 |     0.96 |       0.81 |  40152    |
| Accuracy     |        0.78 |     0.78 |       0.78 |      0.78 |
| Macro Avg    |        0.82 |     0.78 |       0.77 |  80140    |
| Weighted Avg |        0.82 |     0.78 |       0.77 |  80140    |

**Mispredictions (up to 5 FPs and 10 FNs):**

| Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\a\b\so\lv\e=BR`       |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\a\b\so\lv\e=BR`       |            1 |            0 |
| False Negative | `\a\bro\a\dkR>P`        |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\a\b\so\lv\e=BR`       |            1 |            0 |
| False Negative | `\a\bro\a\dkR>P`        |            1 |            0 |
| False Negative | `\a\bo\l\i\s\h5\*T`     |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\a\b\so\lv\e=BR`       |            1 |            0 |
| False Negative | `\a\bro\a\dkR>P`        |            1 |            0 |
| False Negative | `\a\bo\l\i\s\h5\*T`     |            1 |            0 |
| False Negative | `\a\b\sor\b\er\az`      |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\a\b\so\lv\e=BR`       |            1 |            0 |
| False Negative | `\a\bro\a\dkR>P`        |            1 |            0 |
| False Negative | `\a\bo\l\i\s\h5\*T`     |            1 |            0 |
| False Negative | `\a\b\sor\b\er\az`      |            1 |            0 |
| False Negative | `\a\b\s\i\n\t\h\ejM`    |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\a\b\so\lv\e=BR`       |            1 |            0 |
| False Negative | `\a\bro\a\dkR>P`        |            1 |            0 |
| False Negative | `\a\bo\l\i\s\h5\*T`     |            1 |            0 |
| False Negative | `\a\b\sor\b\er\az`      |            1 |            0 |
| False Negative | `\a\b\s\i\n\t\h\ejM`    |            1 |            0 |
| False Negative | `Ac\a\d\e\m\icr\-`      |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\a\b\so\lv\e=BR`       |            1 |            0 |
| False Negative | `\a\bro\a\dkR>P`        |            1 |            0 |
| False Negative | `\a\bo\l\i\s\h5\*T`     |            1 |            0 |
| False Negative | `\a\b\sor\b\er\az`      |            1 |            0 |
| False Negative | `\a\b\s\i\n\t\h\ejM`    |            1 |            0 |
| False Negative | `Ac\a\d\e\m\icr\-`      |            1 |            0 |
| False Negative | `A\aro\n\i\t\ic\*`      |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\a\b\so\lv\e=BR`       |            1 |            0 |
| False Negative | `\a\bro\a\dkR>P`        |            1 |            0 |
| False Negative | `\a\bo\l\i\s\h5\*T`     |            1 |            0 |
| False Negative | `\a\b\sor\b\er\az`      |            1 |            0 |
| False Negative | `\a\b\s\i\n\t\h\ejM`    |            1 |            0 |
| False Negative | `Ac\a\d\e\m\icr\-`      |            1 |            0 |
| False Negative | `A\aro\n\i\t\ic\*`      |            1 |            0 |
| False Negative | `\a\b\l\a\t\iv\ek\}`    |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\a\d\e\no\m\acv\b\n`   |            0 |            1 |
| False Positive | `\acr\i\n\y\l\lkj\h`    |            0 |            1 |
| False Positive | `\ac\a\n\t\h\a9\a\s\df` |            0 |            1 |
| False Positive | `rrrr\a\d\e\lo\po\d`    |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\a\b\so\lv\e=BR`       |            1 |            0 |
| False Negative | `\a\bro\a\dkR>P`        |            1 |            0 |
| False Negative | `\a\bo\l\i\s\h5\*T`     |            1 |            0 |
| False Negative | `\a\b\sor\b\er\az`      |            1 |            0 |
| False Negative | `\a\b\s\i\n\t\h\ejM`    |            1 |            0 |
| False Negative | `Ac\a\d\e\m\icr\-`      |            1 |            0 |
| False Negative | `A\aro\n\i\t\ic\*`      |            1 |            0 |
| False Negative | `\a\b\l\a\t\iv\ek\}`    |            1 |            0 |
| False Negative | `\a\b\s\e\n\t\er\(S`    |            1 |            0 |

Processing with password_validator:   0%|          | 0/80140 [00:00<?, ?it/s]

## Evaluating Method: **password_validator**

**Confusion Matrix:**

|          |   Predicted 0 |   Predicted 1 |
|:---------|--------------:|--------------:|
| Actual 0 |         34724 |          5264 |
| Actual 1 |         14916 |         25236 |

**Classification Report:**

|              |   precision |   recall |   f1-score |   support |
|:-------------|------------:|---------:|-----------:|----------:|
| Class 0      |        0.7  |     0.87 |       0.77 |  39988    |
| Class 1      |        0.83 |     0.63 |       0.71 |  40152    |
| Accuracy     |        0.75 |     0.75 |       0.75 |      0.75 |
| Macro Avg    |        0.76 |     0.75 |       0.74 |  80140    |
| Weighted Avg |        0.76 |     0.75 |       0.74 |  80140    |

**Mispredictions (up to 5 FPs and 10 FNs):**

| Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\ac\h\i\n\g\_fSS`      |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\ac\h\i\n\g\_fSS`      |            1 |            0 |
| False Negative | `\ac\l\i\n\a\l|=x\$`    |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\ac\h\i\n\g\_fSS`      |            1 |            0 |
| False Negative | `\ac\l\i\n\a\l|=x\$`    |            1 |            0 |
| False Negative | `CR\-|r\*qzNS`          |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\ac\h\i\n\g\_fSS`      |            1 |            0 |
| False Negative | `\ac\l\i\n\a\l|=x\$`    |            1 |            0 |
| False Negative | `CR\-|r\*qzNS`          |            1 |            0 |
| False Negative | `\a\b\sco\n\s\aW\#`     |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\ac\h\i\n\g\_fSS`      |            1 |            0 |
| False Negative | `\ac\l\i\n\a\l|=x\$`    |            1 |            0 |
| False Negative | `CR\-|r\*qzNS`          |            1 |            0 |
| False Negative | `\a\b\sco\n\s\aW\#`     |            1 |            0 |
| False Negative | `\ac\h\i\n\gZu\)C`      |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\ac\h\i\n\g\_fSS`      |            1 |            0 |
| False Negative | `\ac\l\i\n\a\l|=x\$`    |            1 |            0 |
| False Negative | `CR\-|r\*qzNS`          |            1 |            0 |
| False Negative | `\a\b\sco\n\s\aW\#`     |            1 |            0 |
| False Negative | `\ac\h\i\n\gZu\)C`      |            1 |            0 |
| False Negative | `[\gE\m\acc\e\n\tor`    |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\ac\h\i\n\g\_fSS`      |            1 |            0 |
| False Negative | `\ac\l\i\n\a\l|=x\$`    |            1 |            0 |
| False Negative | `CR\-|r\*qzNS`          |            1 |            0 |
| False Negative | `\a\b\sco\n\s\aW\#`     |            1 |            0 |
| False Negative | `\ac\h\i\n\gZu\)C`      |            1 |            0 |
| False Negative | `[\gE\m\acc\e\n\tor`    |            1 |            0 |
| False Negative | `Ac\e\m\e\t\ic\ixR`     |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\ac\h\i\n\g\_fSS`      |            1 |            0 |
| False Negative | `\ac\l\i\n\a\l|=x\$`    |            1 |            0 |
| False Negative | `CR\-|r\*qzNS`          |            1 |            0 |
| False Negative | `\a\b\sco\n\s\aW\#`     |            1 |            0 |
| False Negative | `\ac\h\i\n\gZu\)C`      |            1 |            0 |
| False Negative | `[\gE\m\acc\e\n\tor`    |            1 |            0 |
| False Negative | `Ac\e\m\e\t\ic\ixR`     |            1 |            0 |
| False Negative | `v|\h\d\a\b\sco\n\d\er` |            1 |            0 || Type           | Password                |   True Label |   Prediction |
|:---------------|:------------------------|-------------:|-------------:|
| False Positive | `1111\ac\i\no\sE`       |            0 |            1 |
| False Positive | `\aB\l\a\t\io\n0000`    |            0 |            1 |
| False Positive | `\acroM\i\a\l3`         |            0 |            1 |
| False Positive | `\a\a\a\a7\a\eF\au\l\d` |            0 |            1 |
| False Positive | `\a\b\y\s\mA\l6666`     |            0 |            1 |
| False Negative | `\acc\e\p\t\e\dM|`      |            1 |            0 |
| False Negative | `\ac\h\i\n\g\_fSS`      |            1 |            0 |
| False Negative | `\ac\l\i\n\a\l|=x\$`    |            1 |            0 |
| False Negative | `CR\-|r\*qzNS`          |            1 |            0 |
| False Negative | `\a\b\sco\n\s\aW\#`     |            1 |            0 |
| False Negative | `\ac\h\i\n\gZu\)C`      |            1 |            0 |
| False Negative | `[\gE\m\acc\e\n\tor`    |            1 |            0 |
| False Negative | `Ac\e\m\e\t\ic\ixR`     |            1 |            0 |
| False Negative | `v|\h\d\a\b\sco\n\d\er` |            1 |            0 |
| False Negative | `Ao\n<@\hok;o`          |            1 |            0 |

Processing with password_strength:   0%|          | 0/80140 [00:00<?, ?it/s]

KeyboardInterrupt: 

# Password Complexity Estimators Evaluation Summary

## 1. zxcvbn

**Accuracy:** 78%

**Strengths:**
- High recall (96%) for detecting secure passwords.
- Excellent precision (93%) for identifying insecure passwords.

**Weaknesses:**
- Low recall (60%) for insecure passwords, resulting in many false positives.
- Tends to misclassify passwords with predictable patterns (e.g., `adhamantqweRty`) as secure.

## 2. password_validator

**Accuracy:** 75%

**Strengths:**
- High recall (87%) for insecure passwords.
- Good precision (83%) when identifying strong passwords.

**Weaknesses:**
- Low recall (63%) for secure passwords, leading to high false negatives.
- Struggles with complex passwords containing special characters or unusual structures (e.g., `abutter/ib`, `accession<`).

## 3. password_strength

**Accuracy:** 75%

**Strengths:**
- Perfect recall (100%) for insecure passwords.
- Excellent precision (100%) for secure passwords.

**Weaknesses:**
- Poor recall (50%) for secure passwords, causing numerous false negatives.
- Often misclassifies complex passwords with special characters (`accollevY@6`, `woSlwt,59*tb,`) as weak.

## 4. passwordmeter

**Accuracy:** 79%

**Strengths:**
- Perfect precision (100%) in identifying secure passwords; no false positives.
- Perfect recall (100%) for insecure passwords.

**Weaknesses:**
- Moderate recall (59%) for secure passwords, frequently missing unconventional but secure passwords (`AbaditeG7m`, `z>ca<YIxd]U`).

---

## Overall Analysis and Recommendations

**Common Strengths:**
- All estimators effectively identify weak passwords with high reliability.

**Common Weaknesses:**
- Consistently struggle with accurately recognizing strong passwords containing special characters or unusual patterns, resulting in false negatives.

In short, no single estimator stands out as perfect, and most are struggling with passwords that in fact are secure.